[TIMOB-17964] Android 5.0: Add support for Toolbar

ypbnv opened the pull request:
**JIRA:** https://jira.appcelerator.org/browse/TIMOB-17964

Needs to be merged first: 
 - Titanium: https://github.com/appcelerator/titanium_mobile/pull/9027 - it is already in master
 - Alloy PR: https://github.com/appcelerator/alloy/pull/841


**Description:**
 - Added support for Toolbar - as a separate view and as a custom ActionBar.
 - Extracted "makeImageSource" in a new ResourceHelper class to avoid code repetition.

Having a toolbar as an ActionBar requires using a theme without a default one:
 - Theme.Titanium
 - Theme.AppCompat.Translucent.NoTitleBar
 - Theme.AppCompat.Translucent.NoTitleBar.Fullscreen
 - Theme.AppCompat.NoTitleBar
 - Theme.AppCompat.NoTitleBar.Fullscreen

**Very basic test case**
```js
var window = Ti.UI.createWindow();

var toolbar = Ti.UI.createToolbar({
    width: Ti.UI.FILL,
    extendBackground: true,
    top: 0
});

var button = Ti.UI.createButton({
    title: 'Button',
    right: 10,
    top: 0
});

button.addEventListener('click', function(e) {
    Ti.API.info('BackgroundColor is ' + toolbar.getBackgroundColor());
    Ti.API.info('Title is ' + toolbar.getTitle());
    Ti.API.info('Subtitle is ' + toolbar.getSubtitle());
    Ti.API.info('Logo is ' + JSON.stringify(toolbar.getLogo()));
    Ti.API.info('Overflow icon is ' + JSON.stringify(toolbar.getOverflowIcon()));
    Ti.API.info('Navigation icon is ' + JSON.stringify(toolbar.getNavigationIcon()));
  
    toolbar.showOverflowMenu();
  
    setTimeout(function() {
        Ti.API.info('Is overflow menu showed? - ' + toolbar.isOverflowMenuShowing());
      
        toolbar.hideOverflowMenu();
        toolbar.setLogo('logo2.jpg');
        toolbar.collapseActionView();
        toolbar.dismissPopupMenus();
        toolbar.setBackgroundColor('red');
        toolbar.setTitle('New Title');
        toolbar.setTitleTextColor('orange');
        toolbar.setSubtitle('New Subtitle');
        toolbar.setSubtitleTextColor('blue');
      
        Ti.API.info('Content inset end is ' + toolbar.getContentInsetEnd());
        Ti.API.info('Content inset end with actions is ' + toolbar.getContentInsetEndWithActions());
        Ti.API.info('Content inset left is ' + toolbar.getContentInsetLeft());
        Ti.API.info('Content inset right is ' + toolbar.getContentInsetRight());
        Ti.API.info('Content inset start is ' + toolbar.getContentInsetStart());
        Ti.API.info('Content inset start with navigation is ' + toolbar.getContentInsetStartWithNavigation());
    }, 3000);
});

toolbar.setItems([button]);
window.add(toolbar);

window.activity.onCreateOptionsMenu = function(e) {
    var menu = e.menu;
    
    var menuItem = menu.add({
        title: 'Item 1',
        showAsAction: Ti.Android.SHOW_AS_ACTION_NEVER
    });
    
    menuItem.addEventListener('click', function(e) {
        Ti.API.debug('Item 1 was clicked');
    });
    
    var menuItem2 = menu.add({
        title: 'Item 2',
        showAsAction: Ti.Android.SHOW_AS_ACTION_NEVER
    });
    
    var menuItem3 = menu.add({
        title: 'Item 3',
        showAsAction: Ti.Android.SHOW_AS_ACTION_NEVER
    });
    
    var menuItem4 = menu.add({
        title: 'Item 4',
        showAsAction: Ti.Android.SHOW_AS_ACTION_NEVER
    });
};

window.activity.supportToolbar = toolbar;

window.addEventListener('open', function() {
    toolbar.setBackgroundColor('blue');
    toolbar.setTitle('Title');
    toolbar.setTitleTextColor('red');
    toolbar.setSubtitle('Subtitle');
    toolbar.setSubtitleTextColor('green');
    toolbar.setLogo('logo.png');
    toolbar.setNavigationIcon('logo.png');
    toolbar.setOverflowIcon('logo.png');
    //toolbar.setContentInsetEndWithActions(10);
    //toolbar.setContentInsetStartWithNavigation(11);
    //toolbar.setContentInsetsAbsolute(12,12);
    toolbar.setContentInsetsRelative(13, 13);
});

window.open();
```

**Alloy test:**

```
<Alloy>
    <Window title="My Test App" backgroundColor="gray" customToolbar="toolbar">
        <Toolbar
                 id="toolbar"
                 title="MyMenu"
                 subtitle="Subtitle"
                 width="Ti.UI.FILL"
                 top=0
                 barColor="#639851"
                 displayHomeAsUp="true"
                 homeButtonEnabled="true"
                 overflowIcon="logo.png">
             <Items>
                <Button id="ok" title="OK"/>
                <Button id="cancel" title="Cancel"/>
            </Items>
        </Toolbar>
        <Menu>
            <MenuItem id="item1" title="Settings" showAsAction="Ti.Android.SHOW_AS_ACTION_NEVER"/>
            <MenuItem id="item2" title="Search" showAsAction="Ti.Android.SHOW_AS_ACTION_NEVER"/>
        </Menu> 
    </Window>
</Alloy>
```

Documentation is on the way.

Files for the test case:
![logo](https://user-images.githubusercontent.com/13927169/27230773-116c1aa0-52b9-11e7-8bde-62572b2988e2.png)
![logo2](https://user-images.githubusercontent.com/13927169/27230776-153c5032-52b9-11e7-9d93-18b293856769.jpg)

ypbnv committed:
[TIMOB-17964] Android 5.0: Add support for Toolbar
Initial commit - adds Toolbar proxy with basic functionality.

ypbnv committed:
Remove wildcard import.

ypbnv committed:
Add a missing import.

ypbnv committed:
Add docs.

hansemannn commented on apidoc/Titanium/Android/Activity.yml:
```diff
@@ -193,6 +193,12 @@ methods:
+    summary: Sets a toolbar instance to be used as an ActionBar.
+    parameters:
+      - name: toolbar
+        type: Titanium.UI.Android.Toolbar  
```
The setter/getter are generated by the apidoc-generator automatically, you don't need to specify it.

ypbnv replied:
I haven't described that as I should. I will add a more detailed description.
The Activity class does not handle property changes. So this method is used if you want to set a Toolbar as ActionBar after the instance is created. So technically currently this is not a setter for the supportToolbar (maybe I should assign it the toolbar reference).

hansemannn commented on apidoc/Titanium/Android/Activity.yml:
```diff
@@ -435,6 +441,10 @@ properties:
 
+  - name: supportToolbar
+    summary: Toolbar instance that serves as ActionBar
+    type: Titanium.UI.Android.Toolbar
```
Also add the `since: "6.2.0"` key :-)

ypbnv replied:
My bad. Will add it.

hansemannn commented on apidoc/Titanium/UI/Android/Toolbar.yml:
```diff
@@ -0,0 +1,255 @@
+  - name: getContentInsetEnd
+    summary: Returns the margin at the toolbar's content end.
+    returns:
+        type: Number
```
Same as above: No manual setter / getter needed.

ypbnv replied:
I haven't added the content insets as properties, so these are directly setting/getting values to/from the native view. I wanted to keep the properties the same as the iOS ones (except borders) and just have Android only methods.

hansemannn replied:
Not sure what you mean. iOS does not expose the property, and the getter / setter are generated by the build-script, so when you explicitly write them, they will likely be generated twice. Example: [This line](https://github.com/appcelerator/titanium_mobile/blob/master/apidoc/Titanium/UI/View.yml#L1053) in the yml-files of `Ti.UI.View` will produce [this property](http://docs.appcelerator.com/platform/latest/#!/api/Titanium.UI.View-property-backgroundColor), but also the [getter](http://docs.appcelerator.com/platform/latest/#!/api/Titanium.UI.View-method-getBackgroundColor) and [setter](http://docs.appcelerator.com/platform/latest/#!/api/Titanium.UI.View-method-setBackgroundColor) of it.

ypbnv replied:
Yes, I got that. 
There is no property "contentInsetEnd" in the Toolbar.yml. This is because there is no such property in the ToolbarProxy in Android.

hansemannn replied:
Alrighty. Just thought that people would still like the `contentInsetEnd` property, but let's keep it then.

ypbnv committed:
Add missing 'since' key. Add a description.

garymathews commented on android/modules/android/src/java/ti/modules/titanium/android/notificationmanager/BigPictureStyleProxy.java:
```diff
@@ -93,7 +83,7 @@ public void setBigLargeIcon(Object icon)
 	public void setBigPicture(Object picture)
 	{
-		TiDrawableReference source = makeImageSource(picture);
+		TiDrawableReference source = ResourceHelper.getInstance().makeImageSource(this,picture);
```
Remember to always space your arguments `(this, picture)`

garymathews commented on android/modules/filesystem/src/java/ti/modules/titanium/filesystem/ResourceHelper.java:
```diff
@@ -0,0 +1,28 @@
+package ti.modules.titanium.filesystem;
```
Maybe this should go in `org.appcelerator.titanium.util` with our other helpers?

ypbnv replied:
That will create a circular dependency between 'titanium' and 'titanium-filesystem' modules.

ypbnv replied:
That's way better. I didn't realize I could use it.

garymathews commented on android/titanium/src/java/org/appcelerator/titanium/proxy/ActivityProxy.java:
```diff
@@ -279,6 +283,13 @@ public void invalidateOptionsMenu()
+	@Kroll.method
+	public void setSupportActionBar(TiToolbarProxy tiToolbarProxy) {
+		TiBaseActivity activity = (TiBaseActivity) getWrappedActivity();
+		if (activity != null)
```
Brackets

```Java
if (activity != null) {
    activity.setSupportActionBar((Toolbar)tiToolbarProxy.getToolbarInstance());
}
```

garymathews commented on apidoc/Titanium/Android/Activity.yml:
```diff
@@ -193,6 +193,16 @@ methods:
+  - name: setSupportActionBar
+    summary: Sets a toolbar instance to be used as an ActionBar.
+    description: |
+        This methos is used if you want to add a Toolbar as an ActionBar after the Activity has been created.
```
Typo, `methos`

garymathews reviewed (changes requested):
CR: **FAIL**

ypbnv committed:
Fixes code and a typo.

ypbnv commented:
@garymathews Fixed typo and code.

ypbnv committed:
Spacing fix.

jquick-axway reviewed (commented):
*Question:*
Since you're using "android.support.v7.widget.Toolbar" from the support library, this may mean that your new Toolbar feature is supported on Android 4.0 (API Level 14), which is the minimum API Level Titanium supports. Google's docs kind of suggest this here...
https://developer.android.com/training/appbar/setting-up.html

Can you please verify this please?

If this is true, then your doc changes no longer need to state that Android 5.0 is the min version needed anymore.

ypbnv commented on android/titanium/src/java/org/appcelerator/titanium/proxy/ActivityProxy.java:
```diff
@@ -53,7 +56,9 @@
 	private KrollFunction resultCallback;
-	
+
+	public TiToolbarProxy toolbarProxy;
```
Removed.

ypbnv commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiToolbar.java:
```diff
@@ -0,0 +1,605 @@
+	 */
+	public TiToolbar(TiViewProxy proxy) {
+		super(proxy);
+		toolbar = new Toolbar(TiApplication.getAppCurrentActivity());
```
I guess proxy.getActivity() should guarantee us that the proper one is passed?

ypbnv committed:
Remove ResourceHelper.
Guard for passing the correct activity during creation.
Guard for null parameter.
Remove redundant variable.

ypbnv commented:
@jquick-axway Updated the PR.
As for the minimum version I suppose you refer to this line
`osver: {android: {min: "21"}}`
in Toolbar.yml ?
If that's the case the requirement applies only for 'extendBackground' property because it depends on making the StatusBar transparent. And that is supported only below API 21 (I may be wrong here).

Otherwise the Toolbar is working fine on every version down to API 16. Which is the minimum Android version required for SDK 6.0.0+, according to the compatabilty matrix. I haven't tested it on API 15 and 14.

jquick-axway commented on apidoc/Titanium/UI/Android/Toolbar.yml:
```diff
@@ -0,0 +1,255 @@
+    A toolbar can be positioned everywhere in the the layout as a [View](Titanium.UI.View). It can also be used as an ActionBar 
+    for [activities](Titanium.Android.Activity). This allows a toolbar to inherit [ActionBar's](Titanium.Android.ActionBar) methods,
+    properties and events and provide a better customization of this design element. For example you can provide your own images
+    to be used as navigation button icon, nverflow menu icon and logo. In order to do that an application must be using Theme.AppCompat.NoTitleBar.
```
Typo: Change "nverflow" to "overflow".

jquick-axway replied:
Also, the following needs a comma. Please change...
`In order to do that an application...`
...to...
`In order to do that, an application...`

jquick-axway commented on apidoc/Titanium/UI/Android/Toolbar.yml:
```diff
@@ -0,0 +1,255 @@
+  - name: extendBackground
+    summary: If `true`, the background of the toolbar extends upwards by the height of StatusBar for current device.
+    description: |
+        This property is Numberended to be used when the toolbar is used as an ActionBar. It allows the user to specify
```
I'm not sure what you mean by "Numberended".

Also, the "extendBackground" Java code appears to auto-size and auto-position the toolbar. So, was this description intended for the "translucent" property?

ypbnv replied:
"Numberended" happens when you have written "intended" and you replace all "int" to "Number" to match the documentation format. : )

'extendBackground' shouldn't be doing that - it adds padding at the top equal to the StatusBar height and makes the latter's background transparent. It will do it wherever you place your toolbar and even if it is not passed as an ActionBar. Can you provide me a snippet where that happens?

jquick-axway replied:
Regarding the "extendBackground", I was more confused about the wording in the documentation. I haven't actually tested it.

jquick-axway commented on apidoc/Titanium/UI/Android/Toolbar.yml:
```diff
@@ -0,0 +1,255 @@
+    summary: Returns the source of the navigation image in the format it was set.
+
+  - name: getOverflowIcon
+    summary: Returns the source of the overflow menu image in the format it was set.
```
Perhaps the "return:" documentation for getLogo(), getNavigationIcon(), and getOverflowIcon() should be set to...
`[String, Titanium.Blob, Titanium.Filesystem.File]`

...like how it is for their setters.

Also, perhaps it would be better to document these as properties.

ypbnv replied:
If we document them as properties that would mean that you can access them directly and the widget is listening for their change. This is not the case with the Logo, OverflowIcon, NavigationIcon. They can be changed only through these methods. We can change them to be properties (also the content inset ones) if that will be more convenient from user stand point.

jquick-axway replied:
They don't have to be properties, but ideally they should be accepted as dictionary field settings in your createToolbar() function. The reason is because this is how an Alloy XML file would pass a "Toolbar" element's attributes to the createToolbar() function. And I believe most of a create function's dictionary fields are expected to be the object's properties as well.

@hansemannn, do you have have any advise regarding this?

jquick-axway commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiToolbar.java:
```diff
@@ -0,0 +1,606 @@
+	 * @param value Boolean value to set. True makes the Toolbar's background extend. False - default behaviour.
+	 */
+	private void handleBackgroundExtended(boolean value) {
+		if (value) {
```
Calling `handleBackgroundExtended(false)` will do nothing. So, once you've set it to true, it's locked-in forever. This was the intent, right?

If so, then I would get rid of the "value" argument.

ypbnv replied:
Thanks, I have missed that.

jquick-axway commented:
We should also document an example on how to set up a Toolbar via Alloy and XML, like how we document it for ActionBar. This may be important for developers who wish to migrate their ActionBar Alloy app settings to the new Toolbar way of doing things.
   http://docs.appcelerator.com/platform/latest/#!/api/Titanium.Android.ActionBar

ypbnv commented:
@jquick-axway Alloy itself will need some changes to support the Toolbar on Android. I did some adjustments, but they may not be relevant if we decide to move the widget in Ti.UI for both platforms.

ypbnv committed:
Merge branch 'master' into TIMOB-17964-upgraded

ypbnv committed:
Move to Ti.UI package.
Guard for missing translucency value.

ypbnv commented:
@hansemannn, @jquick-axway, @garymathews Updated the PR.
Alloy PR for the changed namespace: https://github.com/appcelerator/alloy/pull/841
Maybe the best approach for update in the docs would be after we are done on the iOS too - merge the current iOS documentation, describe platform specific properties and add new methods/properties?

hansemannn commented:
Now that we have the Alloy PR up, I can quickly do the iOS part and do the docs there, adding the Android platform there as well. Not sure when QE can validate this PR, but the iOS one will be very straight-forward (this week'ish).

lokeshchdhry commented:
FR Passed.

Toolbar, its properties, methods & events from view seem to work as expected.

You need to have custom theme with no actionbar in the app:
builtin_themes.xml:
```
<?xml version="1.0" encoding="utf-8"?>
<resources xmlns:android="http://schemas.android.com/apk/res/android">
	<style name="NoActionBar" parent="Theme.AppCompat.NoTitleBar"/>
</resources>
```
tiappp.xml:
```
<android xmlns:android="http://schemas.android.com/apk/res/android">
	    <manifest>
	        <application android:theme="@style/NoActionBar"/>
	    </manifest>
</android>
```

Studio Ver:     4.9.1.201707200100
SDK Ver:        7.0.0 local build
OS Ver:         10.12.3
Xcode Ver:      Xcode 8.3.3
Appc NPM:       4.2.9
Appc CLI:       6.2.3
Ti CLI Ver:     5.0.14
Alloy Ver:      1.9.13
Node Ver:       6.10.1
Java Ver:       1.8.0_101
Devices:        ⇨ google Pixel --- Android 7.1.1
                ⇨ google Nexus 5 --- Android 6.0.1

Commit:
Merge branch 'master' into TIMOB-17964-upgraded

Commit:
Merge branch 'master' into TIMOB-17964-upgraded

Commit:
Merge branch 'master' into TIMOB-17964-upgraded

Commit:
Merge branch 'master' into TIMOB-17964-upgraded

garymathews commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiToolbar.java:
```diff
@@ -0,0 +1,608 @@
+		int width = widthDimension != null ? widthDimension.getAsPixels(toolbar):Toolbar.LayoutParams.WRAP_CONTENT;
+		TiDimension heightDimension = source.getLayoutParams().optionHeight;
+		int height = heightDimension != null ? heightDimension.getAsPixels(toolbar):Toolbar.LayoutParams.WRAP_CONTENT;
+		res.setLayoutParams(new Toolbar.LayoutParams(width,height));
```
space after `,`

garymathews commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiToolbar.java:
```diff
@@ -0,0 +1,608 @@
+	 */
+	public void setLogo(Object object) {
+		if (!TiApplication.isUIThread()) {
+			TiMessenger.sendBlockingMainMessage(mainHandler.obtainMessage(TOOLBAR_SET_LOGO),object);
```
space after `,`

there are other instances of this, please check

garymathews commented on android/modules/ui/src/java/ti/modules/titanium/ui/widget/TiToolbar.java:
```diff
@@ -0,0 +1,608 @@
+	}
+
+	/**
+	 *  Handler for translucency change
```
minor, fix spacing after `*` here

garymathews commented on android/titanium/src/java/org/appcelerator/titanium/proxy/ActivityProxy.java:
```diff
@@ -21,9 +22,11 @@
 import android.os.Message;
-import android.support.v7.app.AppCompatActivity;;
+import android.support.v7.app.AppCompatActivity;
+import org.appcelerator.titanium.view.TiUIView;;
```
two `;;` at end of line

garymathews reviewed (changes requested):
see above, also create a `6.2.0` backport

ypbnv committed:
Add whitespaces.
Remove unused import.

garymathews reviewed (approved):
CR: **PASS**
👍

sgtcoolguy committed:
Merge branch 'master' into TIMOB-17964-upgraded

ypbnv committed:
Merge branch 'master' into TIMOB-17964-upgraded

ypbnv committed:
Merge docs with iOS. Remove Android only docs.
Extract some properties as such.
Add examples.

ypbnv committed:
Add indentaion.

ypbnv commented:
@garymathews, @jquick-axway, @hansemannn Update:
I added the changes from Joshua's latest feedback.
Merged the documentation in the proper place and added two cross-platform examples.
Extracted a few more Android properties as Titanium properties.

Commit:
Merge branch 'master' into TIMOB-17964-upgraded

lokeshchdhry commented:
@ypbnv , I merged this by mistake before https://github.com/appcelerator/alloy/pull/841. 
Do you want to revert this ?

ypbnv commented:
@lokeshchdhry It should not create issues, but won't be available for testing through Alloy before the changes from https://github.com/appcelerator/alloy/pull/841 are applied.
